pkg/instancetype/instancetype.go

```diff
@@ -357,6 +365,67 @@ func getInstancetypeAPISpec(obj runtime.Object) (interface{}, error) {
 	}
 }
 
+func checkCPUPreferenceRequirements(instancetypeSpec *instancetypev1alpha2.VirtualMachineInstancetypeSpec, preferenceSpec *instancetypev1alpha2.VirtualMachinePreferenceSpec, vmiSpec *virtv1.VirtualMachineInstanceSpec) (*k8sfield.Path, error) {
+	if instancetypeSpec != nil && instancetypeSpec.CPU.Guest < preferenceSpec.Requirements.CPU.Guest {
+		return k8sfield.NewPath("spec", "instancetype"), fmt.Errorf(insufficientInstanceTypeCPUResourcesErrorFmt, instancetypeSpec.CPU.Guest, preferenceSpec.Requirements.CPU.Guest)
+	}
+	if instancetypeSpec == nil && vmiSpec.Domain.CPU != nil {
+		// FIXME - Move PreferredCPUTopology to a string pointer and default in mutation webhook before removing this duplicate defaulting
+		preferredCPUTopology := instancetypev1alpha2.PreferSockets
+		if preferenceSpec.CPU != nil && preferenceSpec.CPU.PreferredCPUTopology != "" {
+			preferredCPUTopology = preferenceSpec.CPU.PreferredCPUTopology
+		}
+		switch preferredCPUTopology {
+		case instancetypev1alpha2.PreferThreads:
+			if vmiSpec.Domain.CPU.Threads < preferenceSpec.Requirements.CPU.Guest {
+				return k8sfield.NewPath("spec", "template", "spec", "domain", "cpu", "threads"), fmt.Errorf(insufficientVMCPUResourcesErrorFmt, vmiSpec.Domain.CPU.Threads, preferenceSpec.Requirements.CPU.Guest, "threads")
+			}
+		case instancetypev1alpha2.PreferCores:
+			if vmiSpec.Domain.CPU.Cores < preferenceSpec.Requirements.CPU.Guest {
+				return k8sfield.NewPath("spec", "template", "spec", "domain", "cpu", "cores"), fmt.Errorf(insufficientVMCPUResourcesErrorFmt, vmiSpec.Domain.CPU.Cores, preferenceSpec.Requirements.CPU.Guest, "cores")
+			}
+		case instancetypev1alpha2.PreferSockets:
+			if vmiSpec.Domain.CPU.Sockets < preferenceSpec.Requirements.CPU.Guest {
```

Comment:
Does a socket always translate to exactly one CPU? `PreferSockets` mandates a certain amount of sockets, but does it equal the count of CPUs?

Reply:
Yes, it's a direct copy of the value from the instance type at the moment, future policies will likely be different but for now it means this is a simple check.